### ISSUE TITLE
Code Coverage fails for XSLT that calls transform()

### ISSUE THREAD

**Issue by galtm** [xspec/xspec#746]
If the XSLT being tested calls the `transform()` function, XSpec can still work but the code coverage reporting fails. The Java exception mentions the `XSLTCoverageTraceListener` class that is part of this project. Here is a pair of files that reproduce the problem in Oxygen 21.1 using the "XSpec Code Coverage" transformation in #738.

**File code_coverage_java_error.xsl:**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:transform xmlns:xsl="http://www.w3.org/1999/XSL/Transform" 
  xmlns:mf="helper-function"
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  version="3.0">

  <xsl:template name="template-to-call">
    <output-element/>
  </xsl:template>  

  <xsl:function name="mf:call-some-template">
    <xsl:param name="context" as="element()?"/>
    <!--
      Using xsl:call-template like
      <xsl:call-template name="template-to-call"/>
      works with code coverage, but using
      transform() does not.

      Making 'stylesheet-location' point to a copy of this XSLT file
      instead of the same one, or using 'stylesheet-text' instead
      of 'stylesheet-location', does not help.

      Setting 'enable-trace' property (to either true or false)
      in the map of transform() options does not help.
    -->
    <xsl:sequence select="
      transform(map{
      'stylesheet-location': static-base-uri(),
      'source-node': $context,
      'initial-template': QName('','template-to-call')
      })?output
      "/>
  </xsl:function>
  
</xsl:transform>
```

**File code_coverage_java_error.xspec:**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<x:description xmlns:x="http://www.jenitennison.com/xslt/xspec"
  stylesheet="code_coverage_java_error.xsl"
  xmlns:mf="helper-function" xslt-version="3.0">
  
  <!-- Test passes using "Run XSpec Test" scenario, but code coverage scenario fails with this error.
    
run-xslt-test:
     [echo] Running XSLT Tests...
     [echo] Collecting test coverage data...
     [java] ****************************************
     [java] Testing with SAXON EE 9.8.0.12
     [java] controller=net.sf.saxon.Controller@25f7391eTest of function that uses transform()
     [java] 
     [java] controller=net.sf.saxon.Controller@32057e6
     [java] output-element
     [java] java.lang.IllegalStateException: writeEndElement with no matching writeStartElement
     [java] 	at net.sf.saxon.event.StreamWriterToReceiver.writeEndElement(StreamWriterToReceiver.java:446)
     [java] 	at com.jenitennison.xslt.tests.XSLTCoverageTraceListener.close(XSLTCoverageTraceListener.java:111)
     [java] 	at net.sf.saxon.Controller.callTemplate(Controller.java:2664)
     [java] 	at net.sf.saxon.s9api.Xslt30Transformer.callTemplate(Xslt30Transformer.java:750)
     [java] 	at net.sf.saxon.Transform.processFile(Transform.java:1252)
     [java] 	at net.sf.saxon.Transform.doTransform(Transform.java:782)
     [java] 	at net.sf.saxon.Transform.main(Transform.java:81)
     [java] Fatal error during transformation: java.lang.IllegalStateException: writeEndElement with no matching writeStartElement
    -->

  <x:scenario label="Test of function that uses transform()">
    <x:call function="mf:call-some-template">
      <x:param><input-element/></x:param>
    </x:call>
    <x:expect label="output-element" test="/*">
      <output-element/>
    </x:expect>
  </x:scenario>
</x:description>
```

**Comment by AirQuick** [comment]
Also, Code Coverage may be incompatible with `xsl:for-each/@saxon:threads`. (not tested yet)
Probably, the specification of Coverage XML needs some more careful thought.

**Comment by AirQuick** [comment]
Fixed by #1227
